api/v1alpha3/azuremachine_validation.go

```diff
@@ -165,6 +174,26 @@ func ValidateManagedDisk(old, new ManagedDisk, fieldPath *field.Path) field.Erro
 	return allErrs
 }
 
+// ValidateManagedDisks validates updates to Data disks' managed disk field.
+func ValidateManagedDisks(old, new []DataDisk, fieldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(old) != len(new) {
+		allErrs = append(allErrs, field.Invalid(fieldPath, new, "changing data disks after machine creation is not allowed"))
+		return allErrs
+	}
+
+	for i, disk := range new {
+		if disk.ManagedDisk != nil && old[i].ManagedDisk != nil {
+			if errs := ValidateManagedDisk(*old[i].ManagedDisk, *disk.ManagedDisk, fieldPath.Child("managedDisk")); len(errs) > 0 {
```

Comment:
Do we expect old and new disk slices to be in the same order? Maybe map it to dictionary with `NameSuffix` as key, and then check for equality?

Reply:
Good call out. I will double check and use a dictionary